### ISSUE TITLE
Embedded Quads: Should RDF* allow terms to be included in the graph position?

### ISSUE THREAD

**Issue by blake-regalia** [w3c/rdf-star#49]
The purpose of this issue is to consider what are the merits and what are the drawbacks of allowing a term to appear in the graph position of an embedded triple. In other words, should RDF* deal explicitly in triples, or should embedded triples be generalized to embedded quads?

For example:
```
<< :a :b :c :graph >> :p1 :o1 .
```

Now I can certainly imagine legitimate use cases for embedded quads, but have not dedicated the time to really think this over. So perhaps instead, I will just open the discussion with some basic implications and others can chime in whether they see utility or hinderance:
 - if embedded quads generalizes embedded triples, then the semantics of an embedded triple where the graph component is missing needs to be defined. Presumably this would inherit the graph of the surrounding context and fallback to the "DefaultGraph" when the context is not explicit.
 - think about SPARQL* results formats, `"type": "quad", ...` and `<quad>...</quad>` rather than `"type": "triple", ...` and `<triple>...</triple>` for all embedded statements?
 - the equivalent RDF reification does not have a notion of graphs, i.e., an `rdf:Statement` has no such `rdf:graph` relation. This seems to set precedent against allowing a graph component.

**Comment by klinovp** [comment]
I support this and in fact the implementation in Stardog uses annotations on quads, not triples. I mentioned the motivation in https://github.com/w3c/rdf-star/issues/33 but that issue was primarily about PG representation, not named graphs. The main motivation for quads is to allow RDF* graphs like
```
:g { << :a :p :b >> :label "hello" }
```
This is a valid TriG* in Stardog which means that the annotation on `:a :p :b` is a quad in `:g`. For us it's important to be able to store RDF* triples in NGs, just as regular RDF triples. 

But then since Stardog implements the PG semantics the question, of course, becomes: what's the graph for `:a :p :b` (since it's asserted)? We made the simplifying assumption that it's also `:g` so the TriG* snippet above is a sugar for
```
:g { :a :p :b . << :a :p :b >> :label "hello" }
```
That works pretty well for us so far. Every named graph in RDF* is treated like a container for RDF* triples, some of which can be over embedded triples. We can move embedded triples with annotations freely across graphs and that's pretty great. Not allowing people to move some triples to an NG only because they use annotations would be a severe usability limitation.

I can see reasonable arguments that it might be useful to store `:a :p :b` in `:g1` but `<< :a :p :b >> :label "hello"` in `:g2`. We just didn't have enough demand for that use case to extend the syntax even further (e.g. `<< :a :p :b :g1 >> :label "hello"` like you suggested). That'd require similar changes to SPARQL* syntax too.

I can imagine that named graphs could create problems for the reification-based semantics. That has not been a concern for us at all.

**Comment by pchampin** [comment]
Just for the record: I don't think that the requirementrs expressed by @klinovp require to extend the notion of embedded triples into embedded quads. In fact, I believe that embedded quads would create problems with those requirements.

More precisely:
> ```
> :g { << :a :p :b >> :label "hello" }
> ```
> This is a valid TriG* in Stardog which means that the annotation on :a :p :b is a quad in :g. 

Correct; the subject of this quad, though, is an embedded *triple*.

> For us it's important to be able to store RDF* triples in NGs, just as regular RDF triples.

Which the current abstract syntax allows (https://w3c.github.io/rdf-star/rdf-star-cg-spec.html#dfn-dataset).

> But then since Stardog implements the PG semantics the question, of course, becomes: what's the graph for `:a :p :b` (since it's asserted)?

I think the question is incorrectly phrased. By definition, a triple has a subject, a predicate and an object, but no graph. Asking what its graph is makes no sense. The question should be "in which graph is `:a :p :b` asserted?" where `:a :p :b` is still a triple.

Allowing RDF* to have embedded quads would be problematic in PG mode.
```
:g1 { << :a :p :b :g2 >> :label "hello" }
```
In that example, in which graph would `:a :p :b` be asserted? `:g2` only? `:g1` only? both? No solution 

> I  can see reasonable arguments that it might be useful to store `:a :p :b` in `:g1` but `<< :a :p :b >> :label "hello"` in `:g2`. 

This is trivial in SA mode, of course, still with embedded *triples*.
```
:g1 { :a :p :b  }
:g2 { << :a :p :b >> :label "hello" }
```

**Comment by pchampin** [comment]
I understand how this might look like a natural *generalization* of embedded triples, but I think this is a much more disruptive change.

In RDF in general, I don't see quads as a generalization of triples, they have a very different nature.

* As the building block of graphs, triples are covered by  RDF's semantics. Each triple (*s,p,o*) has a meaning, it makes a *statement* ("the thing named *s* is in the relation named *p* with the thing named *o*").
* Quads, on the other hand, make no statement. Depending on the [chosen semantics for datasets][1], a quad (*s,p,o,g*) may or may not entail the triple (*s,p,o*), it may mean that that statement (*s,p,o*) was made at the address *g*, or made about the thing named *g*...

So while I can envision the use-cases that ``<< :a :b :c >> :d :e`` allows to solve, I really have no idea what use I could make of ``<< :a :b :c :d >> :e :f``.

I would go even further: with embedded triples, we have an opportunity to make explicit the relationship between a graph name and the triples in this graph, something like:
```
:g1 :rel1 << :a :b :c >>, << :d :e :f >>.
:g2 :rel2 << :a :b :c >>, << :h :i :j >>.
```
[1]: https://www.w3.org/TR/rdf11-datasets/

**Comment by klinovp** [comment]
Yes, thanks, I meant exactly "in which graph is `:a :p :b` asserted?". So perhaps this issue is not the right place for my comment, sorry about that! I care a lot more about asserting RDF* triples in named graphs than embedded quads.

**Comment by pchampin** [comment]
@rat10 
> > I really have no idea what use I could make of `<< :a :b :c :d >> :e :f`.
> Oh, come on: this is the original provenance usecase.

No, genuinely, I don't. You seem to assume that the 4th element of a quad unambiguously means provenance, but it doesn't. There are many scenarios where it means something else. What exactly, the quad itself does not say.

**Comment by TallTed** [comment]
Potentially helpful background --

* [Original 2008 N-quads proposal](https://web.archive.org/web/20150322024714/http://sw.deri.org/2008/07/n-quads/) (last updated 2012) 
  `<subject> <predicate> <object> <context> .`
* [2014 N-quads 1.1 Recommendation](https://www.w3.org/TR/n-quads/) 
  "subject, predicate, object and graph label of an RDF Triple and the graph it is part of in a dataset"

**Comment by pchampin** [comment]
I know N-quads, but it is only a syntax for putting triples into "boxes", it does not say anything about the meaning of these boxes.

But, granted, I could use embedded quads for saying at least something like that:
```
    <file.nq> :contains << :alice :likes :bob :g >>.
```

**Comment by TallTed** [comment]
@pchampin - N-Quads 1.1 does at least say those boxes are graphs within datasets within the sphere of RDF. That's *something* (rather a lot, from some perspectives, and certinaly much more than the contentless "context" of N-Quads "original draft") about the meaning of the boxes.

`<file.nq> a ex:filesystem_document, <http://www.w3.org/ns/ldp#RDFSource> ; ex:format ex:nquads` would be some meaningful additions to your sample assertion -- not dictated by but fully available within N-Quads.

(Which is not intended to support @rat-10 *assuming* those or any other statements beyond what *is* automatic with N-Quads or any other RDF serialization being contemplated ... which is itself different than *unserialized* RDF.)

**Comment by rat10** [comment]
> @rat10
> 
> > > I really have no idea what use I could make of `<< :a :b :c :d >> :e :f`.
> > > Oh, come on: this is the original provenance usecase.
> 
> No, genuinely, I don't. You seem to assume that the 4th element of a quad unambiguously means provenance, but it doesn't. There are many scenarios where it means something else. What exactly, the quad itself does not say.

No, that's not what I'm assuming. The fourth element in a quad identifies the graph in which the statement occurs and is a prerequisite to make any assertions about a specific triple occurrence. Provenance is just one, albeit quite popular, example for annotations that typically refer to a specific occurrence.
I would not have thought that I have to spell all this out, again, but well.

**Comment by pchampin** [comment]
> The fourth element in a quad identifies the graph in which the statement occurs

It does not *identify* it the same way that an IRI or a literal identifies something in RDF. This is a much looser notion of "identifying", as explained in the introduction of https://www.w3.org/TR/rdf11-datasets/.

> Provenance is just one, albeit quite popular, example for annotations that typically refer to a specific occurrence.

A quad is not a triple occurrence. A quad is an element of a dataset, and since several datasets can contain the same quad, a quad itself has several occurrences. As such, quads don't solve your problem. 

In other words, in one dataset ``:s :p :o :g`` means *"`:s :p :o` occurs in graph `:g`"*, some another one it means *"`:s :p :o` was asserted by author `:g`"*, in yet another one, it means *"`:s :p :o` describes the thing `:g`"*... 

> I would not have thought that I have to spell all this out, again, but well.

ditto

**Comment by rat10** [comment]
@pchampin You keep arguing about technicalities. 
> > The fourth element in a quad identifies the graph in which the statement occurs
> 
> It does not _identify_ it the same way that an IRI or a literal identifies something in RDF. This is a much looser notion of "identifying", as explained in the introduction of https://www.w3.org/TR/rdf11-datasets/.

Okay, it addresses. SPARQL gets by very well without graphs having a formal model-theoretic semantics as it uses the graph name to address the graph. Its FROM syntax guarantees that a graph name works as intended even if the same IRI is used also for some other, very different "thing". The same is true for a `:g` in fourth position. The syntax defines that it points to the graph in which the statement in question occurs. 

> > Provenance is just one, albeit quite popular, example for annotations that typically refer to a specific occurrence.
> 
> A quad is not a triple occurrence. A quad is an element of a dataset, and since several datasets can contain the same quad, a quad itself has several occurrences. As such, quads don't solve your problem.

You are nitpicking on minor issues that are solvable though e.g. some conventions that a "cool graph name" should ba an IRI starting with the address of the DataSet, appended by the name of the graph etc. And even if all that didn't work out and the `:g`could address only Dataset-local graphs it would still be a major improvement. The difference is that it at least tries to address occurrences - and arguably not totaly without success - whereas your proposed semantics completely ingnores this important aspect.

> In other words, in one dataset `:s :p :o :g` means _"`:s :p :o` occurs in graph `:g`"_, some another one it means _"`:s :p :o` was asserted by author `:g`"_, in yet another one, it means _"`:s :p :o` describes the thing `:g`"_...

We don't need to have this discussion about meaning if all we need is to refer to a triple occurrence. As there is no standardized meaning of the fourth element we will have to convey any such meaning through further triples. Addressing works fine nonetheless (and enables those further statements).

**Comment by pchampin** [comment]
@blake-regalia would you object to closing this issue? (or alternatively tagging it as a 'discussion', which does not mandate an immediate change in the report)
